src/algorithms/contractions/vumps_contractions.jl

```diff
@@ -0,0 +1,367 @@
+    indicesBot = [1 + nlegs_tmb, nlegs_tmb, -N, 4 + H + nlegs_tmb, nlegs_tmb - 1]
+    indicesMid = Vector{Vector{Int}}(undef, H)
+    for h in 1:H
+        indicesMid[h] = [
+            3 + 3 * (h + 1), 3 + 3 * h, 2 + 3 * h, -(2 + h), 3 + h + nlegs_tmb, 1 + 3 * h
+        ]
+    end
+    indices = [indicesGL, indicesA, indicesĀ, indicesTop, indicesMid..., indicesBot]
+
+    # record conjflags
+    conjlist = [false, false, true, false, repeat([false], H)..., true]
+
+    # perform contraction, permute to restore partition
+    GL′ = permute(ncon(tensors, indices, conjlist), (Tuple(1:N), (N + 1,)))
+
+    return GL′
+end
+
+# specialize simple case
+function MPSKit.transfer_right(
+    GR::GenericMPSTensor{S,4},
+    O::PEPOSandwich{1},
+    A::GenericMPSTensor{S,4},
+    Ā::GenericMPSTensor{S,4},
+) where {S}
+    return @tensor GR′[-1 -2 -3 -4; -5] :=
+        GR[10 7 4 2; 1] *
+        conj(Ā[-5 9 6 3; 1]) *
+        ket(O)[8; 11 7 9 -2] *
+        only(pepo(O))[5 8; 12 4 6 -3] *
+        conj(bra(O)[5; 13 2 3 -4]) *
+        A[-1 11 12 13; 10]
+end
+
+# general case
+function MPSKit.transfer_right(
+    GR::GenericMPSTensor{S,N},
+    O::PEPOSandwich{H},
+    A::GenericMPSTensor{S,N},
+    Ā::GenericMPSTensor{S,N},
+) where {S,N,H}
+    # sanity check
+    @assert H == N - 3
+
+    # collect tensors in convenient order: env, above, below, top, mid, bot
+    tensors = [GR, A, Ā, ket(O), pepo(O)..., bra(O)]
+
+    # contraction order: GR, A, top, mid..., bot, Ā
+
+    # number of contracted legs for full top-mid-bot stack
+    nlegs_tmb = 5 + 3 * H
+
+    # assign and collect all contraction indices
+    indicesGR = [1, 2, ((1:3:((H + 1) * 3)) .+ 3)..., 2 + nlegs_tmb]
+    indicesA = [-1, 3, ((1:3:((H + 1) * 3)) .+ 4)..., 1]
+    indicesĀ = [-(N + 1), ((2:N) .+ (1 + nlegs_tmb))..., 2 + nlegs_tmb]
+    indicesTop = [6, 3, 2, 3 + nlegs_tmb, -2]
+    indicesBot = [1 + nlegs_tmb, nlegs_tmb, nlegs_tmb - 1, 4 + H + nlegs_tmb, -N]
+    indicesMid = Vector{Vector{Int}}(undef, H)
+    for h in 1:H
+        indicesMid[h] = [
+            3 + 3 * (h + 1), 3 + 3 * h, 2 + 3 * h, 1 + 3 * h, 3 + h + nlegs_tmb, -(2 + h)
+        ]
+    end
+    indices = [indicesGR, indicesA, indicesĀ, indicesTop, indicesMid..., indicesBot]
+
+    # record conjflags
+    conjlist = [false, false, true, false, repeat([false], H)..., true]
+
+    # perform contraction, permute to restore partition
+    GR′ = permute(ncon(tensors, indices, conjlist), (Tuple(1:N), (N + 1,)))
+
+    return GR′
+end
+
+@generated function environment_overlap(
+    GL::GenericMPSTensor{S,N}, GR::GenericMPSTensor{S,N}
+) where {S,N}
+    GL_e = tensorexpr(:GL, (N + 1, (2:N)...), 1)
+    GR_e = tensorexpr(:GR, 1:N, N + 1)
+    return macroexpand(@__MODULE__, :(return o = @tensor $GL_e * $GR_e))
+end
+
+# TODO: properly implement in the most efficient way
+function MPSKit.contract_mpo_expval(
+    AC::GenericMPSTensor{S,N},
+    GL::GenericMPSTensor{S,N},
+    O,
```

Comment:
Is there a reasonable type restriction we could put on this function here? This is quite a strong form of type piracy, in the sense that any new type would get caught by this implementation and PEPSKit owns none of the types. If not, I'm fine leaving it as is but maybe adding a comment about this

Reply:
I'll restrict `O` to a union of the sandwich types to restrict to internal PEPSKit types.